src/frontend/mame/ui/state.cpp

```diff
@@ -0,0 +1,256 @@
+// license:BSD-3-Clause
+// copyright-holders:Nathan Woods
+/***************************************************************************
+
+	ui/state.cpp
+
+	Menus for saving and loading state
+
+***************************************************************************/
+
+#include "emu.h"
+#include "machine.h"
+#include "emuopts.h"
+#include "ui/state.h"
+
+
+namespace ui {
+
+/***************************************************************************
+	BASE CLASS FOR LOAD AND SAVE
+***************************************************************************/
+
+int menu_load_save_state_base::s_last_slot_selected;
+
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
+
+menu_load_save_state_base::menu_load_save_state_base(mame_ui_manager &mui, render_container *container, const char *header, bool disable_not_found_items)
+	: menu(mui, container),
+		m_header(header),
+		m_disable_not_found_items(disable_not_found_items),
+		m_enabled_mask(0)
+{
+}
+
+
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
+
+menu_load_save_state_base::~menu_load_save_state_base()
+{
+	// resume if appropriate (is the destructor really the right place
+	// to do this sort of activity?)
+	if (!m_was_paused)
+		machine().resume();
+}
+
+
+//-------------------------------------------------
+//  itemref_from_slot_number
+//-------------------------------------------------
+
+void *menu_load_save_state_base::itemref_from_slot_number(int slot)
+{
+	return (void *)(size_t)slot;
```

Comment:
Casting from a signed integer to an unsigned integer of a different size has undefined behaviour.  Also, `size_t` is not guaranteed to be pointer-sized (that's what `std::uintptr_t` is for), and C-style casts are less readable than function-style casts.